### ISSUE TITLE
k3d version 3.1.4+ problem with spire

### ISSUE THREAD

**Issue by zoemccormick** [greymatter-io/helm-charts#773]
When running k3d with version 3.1.4 or 3.1.5, Mike and I both separately saw an issue that causes edge to fail on the first request, in the logs:

```bash
[2020-10-23 20:23:51.023][29][error][filter] [:] [filters/gm-jwt-security.go:165] gm-jwt-security filter: DecodeHeaders() - fetchToken(): Bad response code received from call to gm-jwt-security: 503
```

We've tracked this down to have something to do with the spire registrar service.
To verify, exec into the edge pod (when it restarts) and run `curl localhost:8001/certs`, there will only be one certificate, the ingress, and no spiffe certs:

```bash
/app $ curl localhost:8001/certs
{
 "certificates": [
  {
   "ca_cert": [
    {
     "path": "/etc/proxy/tls/edge/ca.crt",
     "serial_number": "1000",
     "subject_alt_names": [],
     "days_until_expiration": "2478",
     "valid_from": "2017-08-09T11:23:08Z",
     "expiration_time": "2027-08-07T11:23:08Z"
    }
   ],
   "cert_chain": [
    {
     "path": "/etc/proxy/tls/edge/server.crt",
     "serial_number": "1000",
     "subject_alt_names": [
      {
       "dns": "greymatter"
      }
     ],
     "days_until_expiration": "3429",
     "valid_from": "2020-03-17T18:36:30Z",
     "expiration_time": "2030-03-15T18:36:30Z"
    }
   ]
  }
 ]
}
```

Edge is not receiving any spiffe certs from sds. This is because there are no entries in the spire server:

```bash
kubectl exec -it server-0 -n spire -c server -- /bin/sh
```

and:

```bash
/opt/spire/bin/spire-server entry show -registrationUDSPath /run/spire/socket/registration.sock
```

and only one entry exists (the registrar itself):

```bash
/opt/spire # /opt/spire/bin/spire-server entry show -registrationUDSPath /run/spire/socket/registration.sock
Found 1 entry
Entry ID      : a4a1ef40-bcfc-45f7-b140-ae9f769a0e7a
SPIFFE ID     : spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node
Parent ID     : spiffe://quickstart.greymatter.io/spire/server
TTL           : default
Selector      : k8s_psat:cluster:quickstart
/opt/spire #
``` 

For comparison, in the spire-registrar service logs in a working installation (k3d 3.1.2 for example), when the agent pods come up we see:

```bash
time="2020-10-23T19:34:54Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node" selectors="k8s:ns:spire,k8s:pod-name:agent-pwgkl" spiffe_id="spiffe://quickstart.greymatter.io/agent"
```

But in this version, these logs never appear.

**Comment by zoemccormick** [comment]
An update on digging into this - the different between what im seeing in the registrar logs in 3.1.2 and in the failing newer versions is a tls error. In 3.1.2, when the agent pods come up:

```bash
time="2020-12-09T22:36:27Z" level=info msg="Dialing server" socket_path=/run/spire/socket/registration.sock
time="2020-12-09T22:36:27Z" level=info msg="Initializing registrar"
time="2020-12-09T22:36:27Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/spire/server" selectors="k8s_psat:cluster:quickstart" spiffe_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node"
time="2020-12-09T22:36:27Z" level=info msg="Serving HTTPS" addr="[::]:8444" skip_client_verification=true
time="2020-12-09T22:37:04Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node" selectors="k8s:ns:spire,k8s:pod-name:agent-zfs29" spiffe_id="spiffe://quickstart.greymatter.io/agent"
time="2020-12-09T22:37:04Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node" selectors="k8s:ns:spire,k8s:pod-name:agent-vxxw5" spiffe_id="spiffe://quickstart.greymatter.io/agent"
time="2020-12-09T22:37:04Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node" selectors="k8s:ns:spire,k8s:pod-name:agent-j7psh" spiffe_id="spiffe://quickstart.greymatter.io/agent"
time="2020-12-09T22:37:04Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node" selectors="k8s:ns:spire,k8s:pod-name:agent-bg5jr" spiffe_id="spiffe://quickstart.greymatter.io/agent"
time="2020-12-09T22:37:04Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node" selectors="k8s:ns:spire,k8s:pod-name:agent-2g4rn" spiffe_id="spiffe://quickstart.greymatter.io/agent"

```

in failing versions:

```bash
time="2020-12-09T22:53:14Z" level=info msg="Dialing server" socket_path=/run/spire/socket/registration.sock
time="2020-12-09T22:53:14Z" level=info msg="Initializing registrar"
time="2020-12-09T22:53:14Z" level=info msg="Created pod entry" parent_id="spiffe://quickstart.greymatter.io/spire/server" selectors="k8s_psat:cluster:quickstart" spiffe_id="spiffe://quickstart.greymatter.io/k8s-workload-registrar/quickstart/node"
time="2020-12-09T22:53:14Z" level=info msg="Serving HTTPS" addr="[::]:8444" skip_client_verification=true
2020/12/09 22:53:26 http: TLS handshake error from 10.42.4.1:55114: remote error: tls: bad certificate
2020/12/09 22:53:26 http: TLS handshake error from 10.42.4.1:17638: remote error: tls: bad certificate
2020/12/09 22:53:26 http: TLS handshake error from 10.42.4.1:51795: remote error: tls: bad certificate
2020/12/09 22:53:26 http: TLS handshake error from 10.42.4.1:48313: remote error: tls: bad certificate
2020/12/09 22:53:26 http: TLS handshake error from 10.42.4.1:53641: remote error: tls: bad certificate
```

I am assuming this has something to do with the way [the validatingwebhookconfiguration](https://github.com/greymatter-io/helm-charts/blob/release-2.3/spire/server/templates/validatingwebhookconfiguration.yaml) is connecting to the registrar service, potentially a change in kubernetes default features/permissions

**Comment by zoemccormick** [comment]
Ok after more digging here is the problem and the fix - in failing k3s versions (ie kubernetes versions 1.19+) I found this in the logs for the kube-apiserver:

```bash
Failed calling webhook, failing open registrar.spire.svc.cluster.local: failed calling webhook "registrar.spire.svc.cluster.local": Post "https://registrar.spire.svc:8444/validate?timeout=30s": x509: certificate relies on legacy Common Name field, use SANs or temporarily enable Common Name matching with GODEBUG=x509ignoreCN=0
```

Indicating that the newer versions of kubernetes use a newer version of Go that deprecated common name - sure enough [this is stated here](https://v1-19.docs.kubernetes.io/docs/setup/release/notes/#api-change-1). The certificate on the validatingWebhookconfiguration was generated by Acert [here](https://github.com/greymatter-io/helm-charts/blob/a8333ea5f6f3fb3a68095a6743c6596e0b25a6ab/ci/scripts/spire-ca.sh#L58). Acert doesn't take a SAN. I added this in [this branch](https://github.com/DecipherNow/acert/tree/san) and regenerated the certificates, adding a SAN this time, and the spire registrar service worked as expected.
